docs/proposals/ldap-group-sync.md

```diff
@@ -0,0 +1,189 @@
+
+OpenShift Resource | LDAP Entry
+------------------ | ----------
+OpenShift Group    | LDAP group
+OpenShift `Group`  | group
+`Group`            |
+___
+
+Linking OpenShift `Groups` to their representations on LDAP servers allows for administrators to have an easier way to manage `Groups` inside OpenShift. This additional functionality will require no additional fields on the OpenShift `Group` object. One can indicate that an OpenShift `Group` mirrors an external representation by setting the `openshift.io/externalRecordProvider` `Annotation` in the `Group` as well as adding a `Label` such as `openshift.io/externalRecordProvider/LDAP` to the `Group`. Other information (e.g. last sync time or resource version) will be stored as `Annotations` on the OpenShift `Group`. Logistical information for the LDAP sync will be stored in an `LDAPSyncOptions` that the administrator will need to provide.
+
+### Use Cases
+* I want to sync all groups I've configured in a sync job (e.g. a specific LDAP server configuration) to OpenShift `Groups`
+* I want to sync specific OpenShift `Groups` with their `externalRecordProviders`
+
+### Data Flow
+* Determine list of LDAP groups to be synced
+* Find LDAP group members
+* Determine OpenShift `Identities` from LDAP group members
+* Determine OpenShift `Users` from OpenShift `Identities`
+* Determine OpenShift `Group` from LDAP group
+* Populate new OpenShift `Group` from resulting OpenShift `UserIdentityMappings` and old OpenShift `Group` fields
+
+The sync command will be exposed as `oadm sync <group(s)> [<group-name>] [filters]` and invoked like `oadm sync group <name>` or `oadm sync groups [--label=selector]`.
+
+## Support for LDAP Group Schemas
+The two main LDAP schemas used to denote groups are 1) as first-class LDAP entries and 2) as LDAP attributes on LDAP first-class entries. In the first schema, members of the LDAP group would be listed in the first-class entry as attributes. In the second schema, LDAP first-class entries for users would list the groups that they are a member of in an attribute. These groups may or may not exist as first-class entries themselves.
+
+## Sync Process
+The `LDAPMemberLister` inside of the `LDAPGroupSyncer` will, for each group that needs to be synced, find the list of entries in the LDAP server that encode the members of the group. The `LDAPCachedMemberLister` has the same functionality for LDAP sync jobs where it is not possible to run an LDAP query to get the members of each group to be synced: when syncing all groups for an LDAP schema where groups are attributes, *etc*. The `LDAPMemberIdentityFactory` creates OpenShift `Identities` for these members. The `UserIdentityMapper` determines the OpenShift `Users` that these `Identities` refer to. Syncing all or some of either LDAP or OpenShift groups is supported.
```

Comment:
I expected to see a `GroupFinder` or some such.  An interface that returns a list of `LDAPGroups`